framework/src/vectorpostprocessors/StatisticsVectorPostprocessor.C

```diff
@@ -0,0 +1,144 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "StatisticsVectorPostprocessor.h"
+
+// MOOSE includes
+#include "MooseVariableField.h"
+#include "ThreadedElementLoopBase.h"
+#include "ThreadedNodeLoop.h"
+
+#include "libmesh/quadrature.h"
+
+registerMooseObject("MooseApp", StatisticsVectorPostprocessor);
+
+template <>
+InputParameters
+validParams<StatisticsVectorPostprocessor>()
+{
+  InputParameters params = validParams<GeneralVectorPostprocessor>();
+  params.addClassDescription("Compute statistical values of a given VectorPostprocessor.  The "
+                             "statistics are computed for each column.");
+
+  params.addParam<VectorPostprocessorName>("vpp",
+                                           "The VectorPostprocessor to compute statistics for.");
+
+  // These are directly numbered to ensure that any changes to this list will not cause a bug
+  // Do not change the numbers here without changing the corresponding code in computeStatVector()
+  MultiMooseEnum stats("min=0 max=1 sum=2 average=3 stddev=4 norm2=5");
+  params.addRequiredParam<MultiMooseEnum>(
+      "stats",
+      stats,
+      "The statistics you would like to compute for each column of the VectorPostprocessor");
+
+  return params;
+}
+
+StatisticsVectorPostprocessor::StatisticsVectorPostprocessor(const InputParameters & parameters)
+  : GeneralVectorPostprocessor(parameters),
+    _vpp_name(getParam<VectorPostprocessorName>("vpp")),
+    _stats(getParam<MultiMooseEnum>("stats")),
+    _stat_type_vector(declareVector("stat_type"))
+{
+}
+
+void
+StatisticsVectorPostprocessor::initialize()
+{
+  const auto & vpp_vectors = _fe_problem.getVectorPostprocessorVectors(_vpp_name);
+
+  // Add vectors for each column, the reason for the extra logic here is that the columns a VPP
+  // produces can change
+  for (const auto & the_pair : vpp_vectors)
+  {
+    const auto & name = the_pair.first;
+
+    if (_stat_vectors.find(name) == _stat_vectors.end())
+      _stat_vectors[name] = &declareVector(name);
+  }
+}
+
+void
+StatisticsVectorPostprocessor::execute()
+{
+  if (processor_id() == 0) // Only compute on processor 0
+  {
+    const auto & vpp_vectors = _fe_problem.getVectorPostprocessorVectors(_vpp_name);
+
+    _stat_type_vector.clear();
+
+    // Add the stat IDs into the first colum
+    for (const auto & stat : _stats)
+    {
+      auto stat_id = stat.id();
+
+      _stat_type_vector.push_back(stat_id);
+    }
+
+    // For each value vector compute the stats
+    for (auto & the_pair : vpp_vectors)
+    {
+      const auto & name = the_pair.first;
+      const auto & values = *the_pair.second.current;
+
+      auto & stat_vector = *_stat_vectors[name];
```

Comment:
Potential accidental insertion... Try to avoid using map [] operators on the RHS - like **ever** :)

Reply:
I've fixed literally dozens of these over the years. Maybe even hundreds. The pedantry is more than worth it. I'm not picking on you, I'm trying to encourage defensive coding. If you want to keep the RHS bracket usage, add a mooseAssert right before it, or at least at the top of the method. Maybe someday we'll refactor and initialize and execute won't appear next to one another in the code.

Reply:
Fair - I will change it.  What is your preferred way here again?  Do we have it written down somewhere (maybe it should be on the code-standards... maybe we should even look for this in precheck!)

Reply:
Almost impossible in a pre-check. Using the bracket operator on the RHS with a vector is common and the only way to get values out. Telling the difference between a map and a vector would require full semantic analysis.  I usually just go with std::map::find(), Robert likes std::map::count() better. If you want to use the brackets, then just put a mooseAssert somewhere before so we can actually guarantee (at least in debug) that the value is there. Finally there is std::map::at() but we've more or less decided against using that because it can throw (which at least finds the problem, but doesn't give useful information when it causes your simulation to crash, especially in parallel).

Reply:
I wonder if it would be possible to write a custom clang plugin that would error if you call std::map::operator[]...

Anyway - I changed it...